Migration guide for Unified Recorder (#19210)
Azure/azure-sdk-for-js: sdk/test-utils/recorder-new/MIGRATION.md

# Migration Guide

> 

Reviewer: Is there an explanation somewhere on why the new recorder is called "Unified"?

Reviewer 2: I believe it’s called unified because there’s a single provider for all languages. There could be a better name though.

Reviewer: I am not questioning the name, just wonder whether we should have some more info here or linking to other docs.

Author: I can definitely add a short thing here about it, but this shouldn't be the only place I think.

> This document outlines key differences between the legacy recorder and the new Unified Recorder client.

Reviewer 3: Please mention the docker prerequisite

Author: It's mentioned, but I'll pull it up: https://github.com/Azure/azure-sdk-for-js/blob/4264ca391c83abbd46b9fff43397686b4a776d02/sdk/test-utils/recorder-new/MIGRATION.md?plain=1#L19

## Installing the Unified Recorder

The first step is to install the unified recorder client library in your package by running

```bash
$ rush add --dev -p @azure-tools/test-recorder-new
```

> You will use this library to control the recorder from your tests. The API is somewhat similar to the legacy recorder, albeit with some important differences. These will be discussed below.

Reviewer 2: ```suggestion
You will use this library to control the recorder from your tests. The API is similar to the legacy recorder. Differences will be discussed below.
```

This sentence sounds doubtful. Here’s an idea, but there are other ways to make it sound more secure, of course.

**⚠️ Important:** Do not uninstall the old recorder (`@azure-tools/test-recorder`) just yet! At this time, there are still utilities from the legacy recorder package that have not been migrated to the new package.

## Changes to NPM scripts

> For the unified recorder client library to work, the [test proxy server] must be active while you are running your tests. Helpers have been added to the `dev-tool` package which manage starting and stopping the test proxy server before and after your tests are run. These helpers use the test proxy server Docker image. Therefore, Docker must be installed in your development environment.

Reviewer 2: `[test proxy server]`? 🤔

Author: It's a link.

> The script defined in `package.json` to run the browser tests should be replaced with `dev-tool run test:browser`, and the script to run the Node tests should look something like `dev tool run test:node-ts-input --mocha=\"--timeout 1200000 'test/**/*.spec.ts'`. In the case of the Node tests, the `--mocha="..."` parameter can be used to specify arguments to be passed through to Mocha. For example:

Reviewer 2: What about a diff?

> The script defined in `package.json` to run the browser tests should be replaced with `dev-tool run test:browser`, and the script to run the Node tests should look something like `dev tool run test:node-ts-input --mocha=\"--timeout 1200000 'test/**/*.spec.ts'`. In the case of the Node tests, the `--mocha="..."` parameter can be used to specify arguments to be passed through to Mocha. For example:

Reviewer 2: This paragraph seems redundant given the table below. Should it focus instead on introducing the table below? It could provide some perspective on how to read the table.

(The table is very readable, but this paragraph could be more useful to the table below anyway)

| Script              | Before migration                                                                                                                                                                         | After migration                                                                                                                        |
| :------------------ | :--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | :------------------------------------------------------------------------------------------------------------------------------------- |
| `unit-test:browser` | `karma start --single-run`                                                                                                                                                               | `dev-tool run test:browser`                                                                                                            |

> | `unit-test:node`    | `mocha -r esm -r ts-node/register --reporter ../../../common/tools/mocha-multi-reporter.js --timeout 1200000 --full-trace --exclude \"test/**/browser/*.spec.ts\" \"test/**/*.spec.ts\"` | `dev-tool run test:node-ts-input --mocha=\"--timeout 1200000 --full-trace --exclude 'test/**/browser/*.spec.ts' 'test/**/*.spec.ts'\"` |

Reviewer 3: Add all the commands.
To help with a quick comparison.

Reviewer 3: `integration-test` and `test`

Author: Didn't want to get too specific -- I feel like these will differ quite a bit between packages, no?

Reviewer 3: Intention is to show how they are different amongst themselves.
`integration-test:node` vs `unit-test:node` and `test:node`.

I would prefer examples that are not opinionated, but that might be harder for understanding.

By adding all the changes here, at least the users would understand what to do for their commands from seeing the changes here.

Author: At least in the template package, those commands just call `unit-test` after a build. Changes shouldn't be required.

Author: And unrelated, but we'll have to address the code coverage issue (using `nyc`) at some point, since that's not working at the moment. It might require an addition to dev-tool?

Reviewer 3: > And unrelated, but we'll have to address the code coverage issue (using nyc) at some point, since that's not working at the moment. It might require an addition to dev-tool?

Right, I was thinking simply support `integration-test:node-ts-input` and `integration-test:node-js-input` and etc.
But how about just `test:node-ts-input-nyc` and `test:node-ts-input`, similarly for js-input.

Author: I don't think I follow. Feel free to suggest the change you have in mind.

## Initializing the recorder

> The approach taken to initialize the recorder depends on whether the SDK being tested uses Core v1 (`core-http`) or Core v2 (`core-rest-pipeline`). If your SDK is on Core v2, read on. If you're still on Core v1, jump to the section on Core v1 below.

Reviewer 2: Links to the packages will help 😊

Reviewer 2: Link to the Core v1 section will help 🙂

Author: Added links!

### For Core v2 SDKs

The recorder is implemented as a custom policy which should be attached to your client's pipeline. Firstly, initialize the recorder:

```ts
let recorder: TestProxyHttpClient;

/*
 * Note the use of function() instead of the arrow syntax. We need access to `this` so we
 * can pass test information from Mocha to the recorder.
 */
beforeEach(function(this: Context) {
  recorder = new TestProxyHttpClient(this.currentTest);

> });

Reviewer 2: Could it be possible to avoid `this` references and instead pass the concrete parameters to the recorder?

If not now, in the future?

Reviewer 2: (If you like this idea, this could be a separate issue, to avoid blocking this PR)

Author: Can you give an example of what you mean? I'm not sure that it's possible to avoid.

Reviewer 2: Well, the recorder only uses some things from the test context. If they were provided through parameters, it would be more obvious to what’s going on -> easier to debug perhaps? It’s a fundamental change, but for fundamental changes it’s now or never

Author: I think I find just being able to pass in a single `this.currentTest` parameter more appealing, to be honest. Otherwise, every package using the recorder will have to pass in the same multiple parameters in their tests. The current approach reduces the duplication to just the one parameter, so is less error prone. I don't know 🤷

Reviewer 2: Depending on how the properties are used, it could be possible to spread `…this` I wonder.

This is a stretch anyway, I don’t require this change, nor will ask for others to agree with my opinion here 😆.

```

To enable the recorder, you should then initialize your SDK client as normal and use the recorder's `configureClient` method. This method will attach the necessary policies to the client for recording to be enabled.

```ts
const client = /* ... initialize your client as normal ... */;
// recorderHttpPolicy is provided as an export from the test-recorder-new package.
recorder.configureClient(client);
```

### For Core v1 SDKs

The recorder library provides a custom `HttpClient` that is then passed to the SDK. This client needs to be initialized as follows:

```ts

> let recorder: TestProxyHttpClientCoreV1;

Reviewer 2: I’m slightly confused by the name of this variable. The previous recorder was a complex entity that went way beyond an `httpClient`. This variable at first glance seems to be analogous to the previous recorder, but it really intends to be passed as an `httpClient`. What if this variable is called `httpClient`?

Reviewer 2: Or perhaps the HTTP client could be a specific property? Like `recorder.httpClient`?

Reviewer 2: (If you like this idea, this could be a separate issue, to avoid blocking this PR)

Author: I like your second idea, but definitely out of scope here, and I suspect it would be a significant refactor. @HarshaNalluru may also have thoughts.

Reviewer 2: It would have been really hard to bring this up before because it wasn’t as easy to see the recorder in context. Please consider this last phase (before migrating clients to the new recorder) to be a critical step to factor in design feedback. After the packages are migrated, the incentives will be against doing big design changes.

Reviewer 3: This recorder also does more things than just being a httpClient.
Such as dealing with the start, stop, variables, and adding sanitizers.

How about renaming the type to `RecorderClient` and `RecorderClientCoreV1`? - intention is that RecorderClient does more than just being a HttpClient.
The same recorderClient(RecorderClientCoreV1) can be passed around as httpClient for core-v1 packages.

Author: I think that would improve things from where we are at the moment, but I wonder if we can do better. At the moment, I think the responsibilities of `TestProxyHttpClient` (with or without `CoreV1`) can be split into two areas:

1. The part which actually deals with the redirection of the requests (the HttpClient implementation for core v1 and the `configureClient` method for core v2); and
2. The interface that users of the recorder call to control (e.g. start and stop) the recorder.

The redirection part (1) that is used depends on whether the client library is core v1 or core v2, but the interface (2) is the same. If we could separate the two (e.g. by having a `httpClient` field as Daniel suggests), I feel like that's a more natural API.

On the other hand, as I said, that would require quite a big design change, so I am a bit on the fence. That said, if we're going to make changes, there really is no time like the present 🎁

Reviewer 2: I really like those ideas @timovv ! 😱🔥

/*
 * Note the use of function() instead of the arrow syntax. We need access to `this` so we
 * can pass test information from Mocha to the recorder.
 */
beforeEach(function(this: Context) {
  recorder = new TestProxyHttpClientCoreV1(this.currentTest);
});
```

When initialising your client in your test, you should pass in the recorder as follows:

```ts
const client = new MyServiceClient(/* ... insert options here ... */, { httpClient: recorder });
```

This will allow requests to be intercepted and redirected to the proxy tool.

## Starting and stopping the recorder

> The recorder must now be started and stopped explicitly. At the beginning of your test, start the recorder as follows:

Author: Start and stop used to be explicit as well, it's just that `record` has changed to `recorder.start`.

```ts
await recorder.start({
  envSetupForPlayback: {
    TABLES_SAS_CONNECTION_STRING: "fakeConnectionString"
  }
});
```

Under the hood, this is powered by the Unified Recorder's sanitizer functionality.

**⚠️Important:** To access environment variables, you must use the `env` export made available from the **new** recorder. This ensures that environment variables are sourced from the correct location (using `process.env` and `dotenv` in Node, and using `window.__env__` via Mocha in the browser), and also means that the environment variables set in `envSetupForPlayback` are used in playback mode.

## Customizations on recordings

A powerful feature of the legacy recorder was its `customizationsOnRecordings` option, which allowed for arbitrary replacements to be made to recordings. The new recorder's analog to this is the sanitizer functionality.

For a simple find/replace, a `GeneralRegexSanitizer` can be used. For example:

```ts
recorder.addSanitizers({
  generalRegexSanitizers: [
    regex: "find",
    value: "replace"
  ],
});
```

This example would replace all instances of `find` in the recording with `replace`.

> Other sanitizers for more complex use cases are also available.

Author: Should we elaborate here? Or is there somewhere we can link to?

Reviewer 3: We should write docs and elaborate examples for the whole API.
I'll start that.

Reviewer 3: However, I would consider connection string and header removal to be other main common use cases, let's add them too in this document along with the generalRegexSanitizer?

Reviewer 2: > We should write docs and elaborate examples for the whole API.

Let’s break that down into individual tasks so that we all can understand how far we are from where we want to go 😊

## AAD and the new `NoOpCredential`

The new recorder does not record AAD traffic at present. As such, tests with clients using AAD should make use of the new `NoOpCredential` provided by the recorder library when in playback mode. For example:

```ts
const credential = isPlaybackMode()
  ? new NoOpCredential();
  : new ClientSecretCredential(/* ... */);
```

Since AAD traffic is not recorded by the new recorder, there is no longer a need to remove AAD credentials from the recording using a sanitizer.

> ## Browser tests and modifications to Karma configuration

Reviewer 3: Please also mention removing this
![image](https://user-images.githubusercontent.com/10452642/146350260-354cac9c-1fd8-42e6-9a87-c54d5d9176e4.png)

> 

Reviewer 3: Also remove this 
![image](https://user-images.githubusercontent.com/10452642/146350435-da57652e-20de-47af-a002-2a8081395469.png)
and related plugin

When running browser tests, the recorder relies on an environment variable to determine where to save the recordings. Add this snippet to your `karma.conf.js`:

```ts
const { relativeRecordingsPath } = require("@azure-tools/test-recorder-new");

> process.env.RECORDINGS_RELATIVE_PATH = relativeRecordingsPath();

Reviewer 3: Also mention adding it in the env variables list

Initializing the recorder
-------------------------

The approach taken to initialize the recorder depends on whether the SDK being tested uses Core v1 (`core-http`) or Core v2 (`core-rest-pipeline`).

### For Core v1 SDKs

The recorder library provides a custom `HttpClient` that is then passed to the SDK. This client\ needs to be initialized as follows:

```ts
let recorder: TestProxyHttpClientCoreV1;

/* 
 * Note the use of function() instead of the arrow syntax. We need access to `this` so we
 * can pass test information from Mocha to the recorder.
 */
beforeEach(function(this: Context) {
    recorder = new TestProxyHttpClientCoreV1(this.currentTest);
});
```

When initialising your client in your test, you should pass in the recorder as follows:

```ts
const client = new MyServiceClient(/* ... insert options here ... */, { httpClient: recorder });
```

This will allow requests to be intercepted and redirected to the proxy tool.

> ### For Core v2 SDKs

Reviewer 3: And then add a link "Skip to the next section if your library is still using core-http(core-v1)"

> ### For Core v2 SDKs

Author: Done, no explicit clickable link necessary though IMO since the sections are pretty short.

  generalRegexSanitizers: [
    regex: "find",
    value: "replace"
  ],
});
```

This example would replace all instances of `find` in the recording with `replace`.

Other sanitizers for more complex use cases are also available.

AAD and the new `NoOpCredential`
--------------------------------

The new recorder does not record AAD traffic at present. As such, tests with clients using AAD should make use of the new `NoOpCredential` provided by the recorder library when in playback mode. For example:

```ts
const credential = isPlaybackMode() 
  ? new NoOpCredential();
  : new ClientSecretCredential(/* ... */);
```

Since AAD traffic is not recorded by the new recorder, there is no longer a need to remove AAD credentials from the recording using a sanitizer.

Browser tests and modifications to Karma configuration
------------------------------------------------------

When running browser tests, the recorder relies on an environment variable to determine where to save the recordings. Add this snippet to your `karma.conf.js`:

```ts
const { relativeRecordingsPath } = require("@azure-tools/test-recorder-new");

process.env.RECORDINGS_RELATIVE_PATH = relativeRecordingsPath();
```

> 

Reviewer 3: Can you also add dev-tool test commands and show how they need to be changed from the existing ones?

Author: I touch on this in the Changes to NPM Scripts section above but can add more specific examples if needed.

Author: Added a table below. I still think the text adds some context so keeping it as well.

# Migration Guide

> This document outlines key differences between the legacy recorder and the new Unified Recorder client. The Unified Recorder replaces the existing `nock`-based recorder with a solution that uses the language-agnostic [test proxy server].

Reviewer 3: ```suggestion
This document outlines key differences between the legacy recorder and the new Unified Recorder client. The Unified Recorder replaces the existing `nock/nise`-based recorder with a solution that uses the language-agnostic [test proxy server].
```

# Migration Guide

This document outlines key differences between the legacy recorder and the new Unified Recorder client. The Unified Recorder replaces the existing `nock/nise`-based recorder with a solution that uses the language-agnostic [test proxy server].

## Prerequisites

- [Docker] is required, as the [test proxy server] is run in a container during testing.

## Installing the Unified Recorder

The first step is to install the unified recorder client library in your package by running

```bash

> $ rush add --dev -p @azure-tools/test-recorder-new

Reviewer 4: Running this adds the following dependency `"@azure-tools/test-recorder-new": "~1.0.0"` note the `~`. We should add it as a caret dependency?

Reviewer 5: yes, caret should be good

Reviewer 4: Oddly enough, my project doesn't resolve `"@azure-tools/test-recorder-new"`.
I added `"@azure-tools/test-recorder-new": "^1.0.0"` to my dev-dependencies and after `rush update` there are no changes in my lock file

Reviewer 5: Can you post your branch?

Reviewer 5: Probably some messy pnpm-lock file or build outputs?

> $ rush add --dev -p @azure-tools/test-recorder-new

Reviewer 5: can you check the flag to add caret and update this? @timovv

> $ rush add --dev -p @azure-tools/test-recorder-new

Author: ```suggestion
$ rush add --dev --caret -p @azure-tools/test-recorder-new
```

Update your test scripts based on the following examples:

| Script              | Before migration                                                                                                                                                                         | After migration                                                                                                                        |
| :------------------ | :--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | :------------------------------------------------------------------------------------------------------------------------------------- |
| `unit-test:browser` | `karma start --single-run`                                                                                                                                                               | `dev-tool run test:browser`                                                                                                            |

> | `unit-test:node`    | `mocha -r esm -r ts-node/register --reporter ../../../common/tools/mocha-multi-reporter.js --timeout 1200000 --full-trace --exclude \"test/**/browser/*.spec.ts\" \"test/**/*.spec.ts\"` | `dev-tool run test:node-ts-input --mocha=\"--timeout 1200000 --full-trace --exclude 'test/**/browser/*.spec.ts' 'test/**/*.spec.ts'\"` |

Reviewer 5: the command got updated, please update accordingly 😊

  generalRegexSanitizers: [
    {
      regex: "find", // This should be a .NET regular expression as it is passed to the .NET proxy tool
      value: "replace",
    },
    // add additional sanitizers here as required
  ],
});
```

This example would replace all instances of `find` in the recording with `replace`.

### ConnectionStringSanitizer

A `ConnectionStringSanitizer` can be used to strip all occurrences of a connection string from a recording. Its usage is very similar to the `GeneralRegexSanitizer`. For example:

```ts
recorder.addSanitizers({
  connectionStringSanitizers: [
    {
      actualConnString: /* the actual connection string to be replaced, usually passed in as an environment variable */,  
      fakeConnString: /* a mock connection string to replace actualConnString with */,
    },
  ],
});
```

### RemoveHeaderSanitizer

`RemoveHeaderSanitizer` can be used to remove specific headers from the recordings as follows:

```ts
recorder.addSanitizers({
  removeHeaderSanitizer: {
    headersForRemoval: ["Header1", "Header2", /* ... */]
  }
});
```

Other sanitizers for more complex use cases are also available.

## AAD and the new `NoOpCredential`

> The new recorder does not record AAD traffic at present. As such, tests with clients using AAD should make use of the new `NoOpCredential` provided by the `@azure-tools/test-credential` when in playback mode. The `createTestCredential` helper also exported by the `test-credential` package will handle switching between NoOpCredential in playback and ClientSecretCredential when recording for you:

Reviewer 5: can mention the rush add command for @azure-tools/test-credential

# Migration Guide

This document outlines key differences between the legacy recorder and the new Unified Recorder client. The Unified Recorder replaces the existing `nock/nise`-based recorder with a solution that uses the language-agnostic [test proxy server].

## Prerequisites

> - [Docker] is required, as the [test proxy server] is run in a container during testing.

Reviewer 4: I'm running within WSL and I had to do 2 additional steps to get the proxy-tool working.

1. Start the docker daemon after installing docker `sudo service docker start`
2. Add myself to the docker user group `sudo usermod -aG docker ${USER}` (then log out and in again)

Reviewer 5: @timovv is it a good idea to refer this https://github.com/Azure/azure-sdk-for-js/blob/main/sdk/test-utils/recorder-new/README.md#with-the-docker-run-command ?

Mention something like "For some reason the dev-tool command did not work for you or the internal docker run step is failing.. try running docker run command manually to see what's going wrong and reach out to us if you're blocked"

Author: > Mention something like "For some reason the dev-tool command did not work for you or the internal docker run step is failing.. try running docker run command manually to see what's going wrong and reach out to us if you're blocked"

Worth mentioning something like that, sure.

Regarding your issue, @joheredi, do you think it would suffice to just say 'ensure the Docker daemon is running and you have permission to use it', which covers the basics -- the actual steps you'd have to take would vary a fair bit depending on what environment you're on, so am hesitant to get too specific.

Reviewer 5: @timovv I was thinking in similar lines, maybe mention Jose's steps in the troubleshooting guide? at the end.

Author: It would also be helpful if dev-tool gave a helpful error message when the test proxy fails to start due to Docker issues, since at the moment it just exits and doesn't print anything useful. I'll raise an issue.

Author: Added a quick note on this in the prerequisites section

Update your test scripts based on the following examples:

| Script              | Before migration                                                                                                                                                                         | After migration                                                                                                               |
| :------------------ | :--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | :---------------------------------------------------------------------------------------------------------------------------- |
| `unit-test:browser` | `karma start --single-run`                                                                                                                                                               | `dev-tool run test:browser`                                                                                                   |

> | `unit-test:node`    | `mocha -r esm -r ts-node/register --reporter ../../../common/tools/mocha-multi-reporter.js --timeout 1200000 --full-trace --exclude \"test/**/browser/*.spec.ts\" \"test/**/*.spec.ts\"` | `dev-tool run test:node-ts-input -- --timeout 1200000 --full-trace --exclude 'test/**/browser/*.spec.ts' 'test/**/*.spec.ts'` |

Reviewer 5: let's mention both `node-ts-input` and `node-js-input`

```ts
const client = /* ... initialize your client as normal ... */;
// recorderHttpPolicy is provided as an export from the test-recorder-new package.
recorder.configureClient(client);
```

### For Core v1 SDKs

The recorder library provides a custom `HttpClient` that is then passed to the SDK. This client needs to be initialized as follows:

```ts
let recorder: Recorder;

/*
 * Note the use of function() instead of the arrow syntax. We need access to `this` so we
 * can pass test information from Mocha to the recorder.
 */
beforeEach(function(this: Context) {
  recorder = new Recorder(this.currentTest);
});
```

When initialising your client in your test, you should pass in the recorder as follows:

```ts
const client = new MyServiceClient(
  /* ... insert options here ... */,
  recorder.configureClientOptionsCoreV1({ /* any additional options to pass through */ }),
);
```

This will allow requests to be intercepted and redirected to the proxy tool.

## Starting and stopping the recorder

> The way that the recorder is started and stopped has changed slightly. At the beginning of your test, start the recorder as follows:

Reviewer 5: mention that this can typically be called from the beforeEach/afterEach block.

  }
});
```

Under the hood, this is powered by the Unified Recorder's sanitizer functionality.

**⚠️Important:** To access environment variables, you must use the `env` export made available from the **new** recorder. This ensures that environment variables are sourced from the correct location (using `process.env` and `dotenv` in Node, and using `window.__env__` via Mocha in the browser), and also means that the environment variables set in `envSetupForPlayback` are used in playback mode.

## Recorder variables

If you want to compute a value at record time and re-use it during playback, the Unified Recorder's variable functionality is for you. This API lets you declare variables which are stored with the recording at record time. During playback, instead of computing the variable afresh, the value will be retrieved from the recording. A use case of this might be to set a value randomly during record time that needs to be the same during playback.

Here is an example:

```ts
const queueName = recorder.variable("queueName", "queue-${Math.floor(Math.random * 1000)}");
// Assume that we have a client that has a createQueue method.

> client.createQueue(queueName);

Reviewer 5: put await

```

In this example, the name of the queue used in the recording is randomized. However, in playback, instead of using the value passed into `recorder.variable`, the value will be retrieved from the recording file. This means that the name of the queue will be consistent between recording and playback.

## Customizations on recordings

A powerful feature of the legacy recorder was its `customizationsOnRecordings` option, which allowed for arbitrary replacements to be made to recordings. The new recorder's analog to this is the sanitizer functionality.

### GeneralRegexSanitizer

For a simple find/replace, a `GeneralRegexSanitizer` can be used. For example:

```ts

> recorder.addSanitizers({

Reviewer 5: async

    {
      actualConnString: /* the actual connection string to be replaced, usually passed in as an environment variable */,
      fakeConnString: /* a mock connection string to replace actualConnString with */,
    },
  ],
});
```

### RemoveHeaderSanitizer

`RemoveHeaderSanitizer` can be used to remove specific headers from the recordings as follows:

```ts
recorder.addSanitizers({
  removeHeaderSanitizer: {
    headersForRemoval: ["Header1", "Header2" /* ... */]
  }
});
```

Other sanitizers for more complex use cases are also available.

## AAD and the new `NoOpCredential`

The new recorder does not record AAD traffic at present. As such, tests with clients using AAD should make use of the new `@azure-tools/test-credential` package, installed as follows:

```bash
$ rush add --dev --caret -p @azure-tools/test-credential
```

This package provides a `NoOpCredential` implementation of `TokenCredential` which makes no network requests, and should be used in playback mode. The provided `createTestCredential` helper will handle switching between NoOpCredential in playback and ClientSecretCredential when recording for you:

```ts
const credential = createTestCredential();

> // You may now pass the created credential to your client.

Reviewer 5: add one more line showing the credential being passed

Author: I left this out as it's fairly trivial and the exact code required to pass the credential will vary depending on the client.

Reviewer 3: `new MyServiceClient(<endpoint>, credential)` should be good, just to remind them what a credential is.

Author: Alright

# Migration Guide

This document outlines key differences between the legacy recorder and the new Unified Recorder client. The Unified Recorder replaces the existing `nock/nise`-based recorder with a solution that uses the language-agnostic [test proxy server].

## Prerequisites

- [Docker] is required, as the [test proxy server] is run in a container during testing. When running the tests, ensure the Docker daemon is running and you have permission to use it. For WSL 2, running `sudo service docker start` and `sudo usermod -aG docker $USER` should be sufficient.

## Installing the Unified Recorder

The first step is to install the unified recorder client library in your package by running

```bash

> $ rush add --dev --caret -p @azure-tools/test-recorder-new

Reviewer 3: recorder 2.0 change is done, please update the document accordingly

Update your test scripts based on the following examples:

| Script                     | Before migration                                                                                                                                                                              | After migration                                                                                                               |
| :------------------------- | :-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | :---------------------------------------------------------------------------------------------------------------------------- |
| `unit-test:browser`        | `karma start --single-run`                                                                                                                                                                    | `dev-tool run test:browser`                                                                                                   |

> | `unit-test:node`           | `mocha -r esm -r ts-node/register --reporter ../../../common/tools/mocha-multi-reporter.js --timeout 1200000 --full-trace --exclude \"test/**/browser/*.spec.ts\" \"test/**/*.spec.ts\"`      | `dev-tool run test:node-ts-input -- --timeout 1200000 --full-trace --exclude 'test/**/browser/*.spec.ts' 'test/**/*.spec.ts'` |
> | `integration-test:browser` | `karma start --single-run`                                                                                                                                                                    | `dev-tool run test:browser`                                                                                                   |
> | `integration-test:node`    | `nyc mocha -r esm --require source-map-support/register --reporter ../../../common/tools/mocha-multi-reporter.js --timeout 5000000 --full-trace \"dist-esm/test/{,!(browser)/**/}*.spec.js\"` | `dev-tool run test:node-js-input -- --timeout 5000000 'dist-esm/test/**/*.spec.js'`                                           |

Reviewer 3: ```suggestion
| `unit-test:node`           | `mocha -r esm -r ts-node/register --reporter ../../../common/tools/mocha-multi-reporter.js --timeout 1200000 --full-trace --exclude \"test/**/browser/*.spec.ts\" \"test/**/*.spec.ts\"`      | `dev-tool run test:node-ts-input -- --timeout 1200000 --exclude 'test/**/browser/*.spec.ts' 'test/**/*.spec.ts'` |
| `integration-test:browser` | `karma start --single-run`                                                                                                                                                                    | `dev-tool run test:browser`                                                                                                   |
| `integration-test:node`    | `nyc mocha -r esm --require source-map-support/register --reporter ../../../common/tools/mocha-multi-reporter.js --timeout 5000000 \"dist-esm/test/{,!(browser)/**/}*.spec.js\"` | `dev-tool run test:node-js-input -- --timeout 5000000 'dist-esm/test/**/*.spec.js'`                                           |
```

full-trace is part of default commands 
![image](https://user-images.githubusercontent.com/10452642/149419471-8c13c645-3e2d-42ea-9d7b-e04021e88302.png)

Author: Those are in the before section and aren't present in the after part

Reviewer 3: oops

Note the difference between the dev-tool `node-ts-input` and `node-js-input` commands:

> - `node-ts-input` runs the tests using `ts-node`, without code coverage.
> - `node-js-input` runs the tests using the built JavaScript output, and generates coverage reporting using `nyc`.

Reviewer 3: ```suggestion
- `node-ts-input` runs the tests using `ts-node`(takes TS files as input), without code coverage.
- `node-js-input` runs the tests using the test (JS) bundle built with rollup, and generates coverage reporting using `nyc`.
```

## Initializing the recorder

The approach taken to initialize the recorder depends on whether the SDK being tested uses Core v1 ([`core-http`]) or Core v2 ([`core-rest-pipeline`]). If your SDK is on Core v2, read on. If you're still on Core v1, [jump to the section on Core v1 below](#for-core-v1-sdks).

### For Core v2 SDKs

The recorder is implemented as a custom policy which should be attached to your client's pipeline. Firstly, initialize the recorder:

```ts
let recorder: Recorder;

/*
 * Note the use of function() instead of the arrow syntax. We need access to `this` so we
 * can pass test information from Mocha to the recorder.
 */
beforeEach(function (this: Context) {
  recorder = new Recorder(this.currentTest);
});
```

> To enable the recorder, you should then initialize your SDK client as normal and use the recorder's `configureClient` method. This method will attach the necessary policies to the client for recording to be enabled.

Reviewer 3: Let's also mention that the client should be exposing the "pipeline" object to get the policy added properly.

Author: It would be a compile error if `pipeline` isn't exposed.

});
```

And at the end of your test (or in an `afterEach` block), stop the recorder:

```ts
await recorder.stop();
```

It is important that `recorder.stop()` is called, or otherwise the next test will throw an error when trying to start the already started recorder. Additionally, it is important that both the `start` and `stop` calls are awaited, for similar reasons.

## Environment variables

In the legacy recorder, the `replaceableVariables` option could be used to specify environment variables that would be replaced in the recording and set during playback. This could be used to ensure that secrets and user-specific options do not appear in the recording body.

The Unified Recorder client provides this functionality through the use of the `envSetupForPlayback` option, which is passed when `recorder.start` is called. Like the legacy recorder, it takes in an object mapping environment variables to what they should be replaced with in the recording. For example:

```ts
await recorder.start({
  envSetupForPlayback: {
    TABLES_SAS_CONNECTION_STRING: "fakeConnectionString",
  },
});
```

Under the hood, this is powered by the Unified Recorder's sanitizer functionality.

> **⚠️Important:** To access environment variables, you must use the `env` export made available from the **new** recorder. This ensures that environment variables are sourced from the correct location (using `process.env` and `dotenv` in Node, and using `window.__env__` via Mocha in the browser), and also means that the environment variables set in `envSetupForPlayback` are used in playback mode.

Reviewer 3: ```suggestion
**⚠️Important:** To access environment variables, you must use the `env` export made available from the **new** recorder. This ensures that environment variables are sourced from the correct location (using `process.env` and `dotenv` in Node, and using `window.__env__` via karma in the browser), and also means that the environment variables set in `envSetupForPlayback` are used in playback mode.
```

## Recorder variables

If you want to compute a value at record time and re-use it during playback, the Unified Recorder's variable functionality is for you. This API lets you declare variables which are stored with the recording at record time. During playback, instead of computing the variable afresh, the value will be retrieved from the recording. A use case of this might be to set a value randomly during record time that needs to be the same during playback.

Here is an example:

```ts
const queueName = recorder.variable("queueName", "queue-${Math.floor(Math.random * 1000)}");
// Assume that we have a client that has a createQueue method.
await client.createQueue(queueName);
```

> In this example, the name of the queue used in the recording is randomized. However, in playback, instead of using the value passed into `recorder.variable`, the value will be retrieved from the recording file. This means that the name of the queue will be consistent between recording and playback.

Reviewer 3: ```suggestion
In this example, the name of the queue used in the recording is randomized. However, in playback, instead of using the value passed into `recorder.variable`, the value will be retrieved from the recording file. This means that the name of the queue will be consistent between record and playback modes.
```

reads better?

Author: No real difference imho, but sure, if you like

```

And then, again in `karma.conf.js`, add the variable to the list of environment variables:

```ts
module.exports = function (config) {
  config.set({
    /* ... */

    envPreprocessor: [
      ,
      /* ... */ "RECORDINGS_RELATIVE_PATH", // Add this!
    ],

    /* ... */
  });
};
```

The following configuration options in `karma.config.js` should be **removed**:

```ts
browserConsoleLogOptions: {
  terminal: !isRecordMode(),
}

/* ... */

jsonToFileReporter: {
  filter: jsonRecordingFilterFunction,  outputPath: ".",
}
```

## Changes to `ci.yml`

You must set the `TestProxy` parameter to `true` to enable the test proxy server in your SDK's `ci.yml` file.

```yaml
# irrelevant sections of ci.yml omitted

extends:
  template: ../../eng/pipelines/templates/stages/archetype-sdk-client.yml
  parameters:
    TestProxy: true # Add me!
```

## Migrating your recordings

Once you have made the necessary code changes, it is time to re-record your tests using the Unified Recorder to complete the migration. To do this, first **delete** the directory containing the old recordings (the `recordings` folder). Then, run your tests with the `TEST_MODE` environment variable to `record`.

> If everything succeeds, the new recordings will be made available in the `recordings` directory. Inspect them to make sure everything looks OK (no secrets present, etc.), and then run the tests in playback mode to ensure everything is passing. If you're running into issues, check out the [Troubleshooting section](#troubleshooting).

Reviewer 3: troubleshooting is empty?

Author: uh oh...

why do I remember writing it if it's not there...

If everything succeeds, the new recordings will be made available in the `recordings` directory. Inspect them to make sure everything looks OK (no secrets present, etc.), and then run the tests in playback mode to ensure everything is passing. If you're running into issues, check out the [Troubleshooting section](#troubleshooting).

## Troubleshooting

If you run into issues while migrating your package, some of the following troubleshooting steps may help:

### Viewing test proxy log output

`dev-tool` by default outputs logs from the test proxy to `test-proxy-output.log` in your package's root directory. These logs can be inspected to see what requests were made to the proxy tool.

### Viewing more detailed logs by running the proxy tool manually

If you desire, you can run the proxy tool docker image manually before running your tests. This allows you to specify a different log level (debug in the below example), allowing for more detailed logs to be viewed. Do this by running:

```bash
docker run -v <your azure-sdk-for-js repository root>:/srv/testproxy -p 5001:5001 -p 5000:5000 -e Logging__LogLevel__Microsoft=Debug azsdkengsys.azurecr.io/engsys/testproxy-lin:latest
```

> Once you've done this, you can run your tests in a separate terminal. `dev-tool` will detect that a test proxy container is already running and will point requests to the Docker container you started.

Reviewer 3: perfect 🎉
